Fix handling of single quotes in RedshiftToS3Operator

ashwaq06 opened the pull request:
closes: #30287

This PR resolves an issue with the RedshiftToS3Operator in Apache Airflow, ensuring correct handling of single quotes in data. It includes refactoring, unit tests, and follows project guidelines. Collaboration with the community has been instrumental in making this contribution.

Commit:
Fix handling of single quotes in RedshiftToS3Operator

ashwaq06 committed:
Merge branch 'main' into main

hussein-awala commented on airflow/providers/amazon/aws/transfers/redshift_to_s3.py:
```diff
@@ -144,7 +144,7 @@ def _build_unload_query(
     ) -> str:
         return f"""
-                    UNLOAD ('{select_query}')
+                    UNLOAD ($${select_query}$$)
```
could you check [this comment](https://github.com/apache/airflow/pull/30300#issuecomment-1486971598)?

ashwaq06 replied:
Thank you for raising this concern. @vincbeck  You are right that using dollar quoting might not be sufficient for queries that already use double quotes. Currently, the suggested change in the code does not account for queries using double quotes, and it may not work as expected in those cases.

To address this issue, one possible solution is to detect if the query contains double quotes and, if it does, avoid using dollar quoting. Instead, the operator can dynamically determine the appropriate quoting method based on the query content.

Additionally, to ensure the robustness of the operator, it is essential to include unit tests that cover various use cases, including queries with single quotes, double quotes, and combinations of both. By adding comprehensive unit tests, we can verify that the `RedshiftToS3 `Operator works correctly with different types of queries.

By implementing these changes and enhancing the operator with proper handling of both single and double quotes, we can provide a more reliable and versatile solution for users dealing with complex queries.

cjames23 replied:
I am not sure this is something worth going to deep on to work around a very limited amount of edge cases. The only use of double quotes would be around column names that contain illegal character sequences. single quotes are for literals which is really going to cover 99% of usage in my opinion. @vincbeck did you have other edge cases in mind from that comment? One other way to approach this would be to simply regex replace any single quote within the string with double single quotes as given in the Redshift Unload docs and continue to wrap the entire query in single quotes. That would cover the double quote usage edge case.

vincbeck replied:
Yeah this solution can work. Or all the way around, you replace `""` with `"` and keep using `$$`. This solution MIGHT be better because, as you pointed out, queries with `""` are rare and as such the replace will be here "just in case". But I definitely think this is worth handling this use case, otherwise your change will be breaking for any query using already `""`.

ashwaq06 committed:
Merge branch 'apache:main' into main

ferruzzi commented on tests/providers/amazon/aws/transfers/test_redshift_to_s3.py:
```diff
@@ -395,3 +395,51 @@ def test_table_unloading_using_redshift_data_api(
+        ]
+
+        # Test a query with single quotes
+        select_query = "SELECT * FROM table WHERE column = 'value with single quotes'"
```
Do we maybe want to parameterize this test to include testing double quotes and maybe missing both types or quotes, just to ensure nothing is breaking?

cjames23 commented on tests/providers/amazon/aws/transfers/test_redshift_to_s3.py:
```diff
@@ -395,3 +395,51 @@ def test_table_unloading_using_redshift_data_api(
+
+        # Assert that the query is wrapped with double dollar sign quotes
+        expected_unload_query = f"""
+            UNLOAD('SELECT * FROM table WHERE column = $$SELECT * FROM table WHERE column = 'value with single quotes'$$')
```
is this a typo? This does not seem to match the above select query being passed in.

ashwaq06 committed:
Merge branch 'apache:main' into main

ashwaq06 committed:
Update test_redshift_to_s3.py

This test case ensures that the RedshiftToS3Operator correctly handles a custom select_query containing single quotes. It validates that the operator generates a proper UNLOAD query that handles single quotes, thereby confirming its functionality in such scenarios.

ashwaq06 committed:
Merge branch 'main' into main

cjames23 commented on tests/providers/amazon/aws/transfers/test_redshift_to_s3.py:
```diff
@@ -216,7 +216,60 @@ def test_custom_select_query_unloading(
+        
+    @pytest.mark.parametrize("table_as_file_name, expected_s3_key", [[True, "key/table_"], [False, "key"]])
+    @mock.patch("airflow.providers.amazon.aws.hooks.s3.S3Hook.get_connection")
+    @mock.patch("airflow.models.connection.Connection")
```
Why are you mocking the connection rather than using the `airflow.utils.db.merge_conn` for your tests? I also think it would be preferable here given what the changes are that tests should use Moto which supports redshift rather than mocking in a way that will not confirm that the redshift engine handles them as expected and does not return any errors.
